### ISSUE TITLE
Screenshot Content Characteristic MTU issue

### ISSUE THREAD

**Issue by thunderluca** [AsteroidOS/asteroid-btsyncd#5]
Hi all,

I'm trying to create a unofficial UWP connector for AsteroidOS smartwatches and I've recently encountered a problem with the (amazing documented) BLE exposed services, in particular with the Screenshot Content Service.

The Request's characteristic operation has always successful result, the application can receive the notifications by Content's characteristic, read the first 4-bytes that determines the image size as int32 but after a few chunks of raw bytes, the Smartwatch (in my case a Sony "tetra" SR50) stop sending data.

The strange fact I observed is sent data is always 10 times lower than expected size.

I think that both SR50 and my phone (Lumia 950XL), that are respectively BT 4.0 and BT 4.1, can't manage a chunk bigger than 23 bytes because of SOC's specification/hardware limitations.

I also noticed in this repo that the max output size sent by AsteroidOS GATT server is actually 200 bytes (I'm not a Qt developer but I read the official Qt documentation for QFile.read method). 

```c
 while (!f.atEnd()) {
        m_value = f.read(200);
        emit valueChanged();
        emitPropertiesChanged();
    }
f.close();
```

I think that is higher of dimensions supported by average consumer devices

**Comment by FlorentRevest** [comment]
Hey @thunderluca !
Thanks a lot for your work on WinSteroid, that sounds super cool! It looks like you are correct, apparently what AsteroidOSSync does is that it sets the MTU itself (from the phone side) to 256 https://github.com/AsteroidOS/AsteroidOSSync/blob/d35e04b104daa159fc8ece977ae16c3b70aacaa1/app/src/main/java/org/asteroidos/sync/services/SynchronizationService.java#L323 and then asteroid-btsyncd just assumes that the MTU is large enough.
Could you try to set the MTU fom the UWP APIs? If that works, we'll need to document that behavior on the BLE profiles doc page.

P.S: Would you be interested in getting WinSteroid upstream? (hosted under https://github.com/AsteroidOS/WinSteroid)

**Comment by thunderluca** [comment]
Hi @FlorentRevest ,

Actually I only tried all tricks to push it with managed code (C#, reflection also failed), I will try a C++ Win32 component to see if it is a really HW limitations of BT 4.0/4.1 controllers when they receive data (or just ARM/x86 Windows bad API set/driver).

About upstreaming WinSteroid it would be really exciting! 
Actually, due to API limitations, the "only" working things I've implemented are live (tile!) battery status and W10 Action Centre notification listener, with applications notifications preferences. 

(Not so much, but I hope to expose something else in the future, like releasing libraries to link Media commands for multimedia apps or Weather apps or exposing sharing syncronization background tasks. 
I'm also trying to find, on Desktop platform, a way to upload Watchfaces via SSH commands as described in your documentation, but that are all ideas).

I'll update this issue trying the Win32 option, but it will take some time (and I'm working, too, so I can develop & mantain WinSteroid in my free time).

**Comment by FlorentRevest** [comment]
I have truly no idea what those tiles are and are made of but it definitely sounds cool... :) Having your code upstream is also great because if someone else wants to pick up your work or help you along the way, it's easier for everyone to collaborate on this.
Let me know what you come up with! :)

**Comment by thunderluca** [comment]
Hello @FlorentRevest,

I'm sorry to say that every attempt to tweak the inbound data transmission rate failed (or, to be clear, I was not able to do it). It seems a driver or hardware limitation of my devices, I also tried with my notebook that has a Qualcomm Wi-Fi/BT NIC but with version 4.1 and seems to have the 23-bytes limitation too.

I don't know what to do in this case, or advising users to use a 4.2+ BT compatibile device to use screenshot feature or anything else.

If there are any guide-lines for Asteroid apps, I'm glad to follow and implement them.

P.S.: I'm ready to upstream my repository whenever you want, I also like the idea to have an upstream repository where merge/PR my changes :-)

**Comment by FlorentRevest** [comment]
Hi @thunderluca ,
Sorry for the delay, life always gets in the way...

That 23-bytes limitation seems to be a real problem. If you feel like recompiling Asteroid for your watch, you can patch asteroid-btsyncd to shorten the 200 bytes long messages and then you'll be able to work on the screenshots profile. If this turns out to be a real BLE limitation and doesn't affect too badly screenshot download time, I'd consider merging your changes in asteroid-btsyncd.

There are no precise guidelines for Asteroid apps but an overall good code quality and good "commit hygiene" (atomic with clear messages) is expected. I'm definitely not an expert of C# but after a quick review it looks like your code is perfectly sane. For the future, you could pick some clearer commit messages but it's already pretty much ok.
That being said, one thing really lacks in your repository: clear licensing. You need to put a LICENSE file at the root of your repository and a clear copyright + license header commentary in each of your .cs file. Feel free to choose whatever open-source license suits you best but make it clear from the very beginning.

I created a repository for you here: https://github.com/AsteroidOS/WinSteroid and you should have write accesses to it so you can just follow the instructions and push your existing repository there.

**Comment by thunderluca** [comment]
@FlorentRevest I beat your delay! :-) (I finished the first beta release of WinSteroid, it passed Microsoft Store certification process yesterday)

I have finally built a tetra asteroidos image with my fork of asteroid-btsyncd, only changing the data size from 200 to 20 bytes, and it works :-/

![screenshot](https://user-images.githubusercontent.com/4592131/41919937-f07a0a4a-795f-11e8-946f-b18e1e0ec0ac.jpg)

So I can finally confirm it is a BT limitations of 4.0/4.1 BLE devices that could not send/receive more than 20 bytes at time.

It didn't take too much time for a 20 bytes transfer at time, from my log it used like 1289 bytes packages to complete it (total size: 25KB, 25755B).

I'm open to any solution to warning user in case both server (watch with AsteroidOS) and Windows client
are not BT 4.2+ compliant, but IMHO asteroid-btsyncd should know his device limitations in some way

**Comment by FlorentRevest** [comment]
Hey @thunderluca,

Great job, congratulations! If you have a Twitter account, could you publish something about WinSteroid with a screenshot attached and send me the link? I'll then retweet it from AsteroidOS' account. If you don't have a Twitter account, could you send me a screenshot of WinSteroid so that I can publish an announcement?

I just changed asteroid-btsyncd to push messages 20bytes by 20bytes. It should be included in the next nightly builds. ;) Let's see if we can find a better workaround later on...

**Comment by thunderluca** [comment]
Ok, I'm the bad screenshots content guy.

I found that screenshots made when AsteroidOS is showing a lightful UI (with a few elements) and image's size is 30KB, the service communicates to the client device without any problem. 

But if you try for example to take a screenshot of Calculator app (50KB circa), the asteroid-btsyncd (or system?) seems to kill the connection after some time and then riconnects, but losing Characteristic registration and, consequently, client's listening session.

I think that with my request of downgrading packets sizes from 200 to 20 I introduced a regression :-/
